torch/csrc/jit/codegen/cuda/executor_utils.cpp

```diff
@@ -672,6 +689,42 @@ NvrtcFunction nvrtcCompile(
 #ifndef NDEBUG
   args.push_back("-lineinfo");
 #endif
+  args.push_back("-G");
+
+  // keeping this outside the loop for lifetime;
+  std::string max_register_usage = "--maxrregcount=";
+  // TODO: double check the math with occupancy header
+  if (opt_block_size.has_value()) {
+    // int num_partition;
+    // cudaOccSubPartitionsPerMultiprocessor(&num_partition, prop);
+    // status = cudaOccRegAllocationGranularity(
+    //     &allocationGranularity,
+    //     properties);
+    int num_partition = 4;
+    int reg_allocation_granularity = 256;
+    int warp_size = 32;
+
+    int num_warps = ceilDiv(opt_block_size.value(), warp_size);
+    int max_warps_per_sm_partition = ceilDiv(num_warps, num_partition);
+
+    int hardware_warps_mem_check = roundUp(num_warps, num_partition);
+    int max_reg_per_warp = prop->regsPerBlock / hardware_warps_mem_check / reg_allocation_granularity * reg_allocation_granularity;
+    int max_register = max_reg_per_warp / warp_size;
+
+    max_register_usage += std::to_string(max_register);
```

Comment:
I'm really confused here why we care about partitions or warps in our register usage check. In my mind we should only be checking 2 things, register per thread is <255 and register per block usage is <65536 (x32-bit registers). I don't understand what else you're trying to check here.

Reply:
I think this is because register partitioning is done by each warp scheduler per SM. The number of warp schedulers is assumed as 4 here, which is I think true except for very old architectures, but we could make it more robust by checking the architecture version number.

I'm assuming Jie is still just prototyping this for now, but more detailed comments would be definitely helpful.

Reply:
Naoya was right, there's a hardware check on register numbers, which assumes register allocations are made to all partitions (warp scheduler) simultaneously, so to simulate the hardware check, we round up the warp number to the number of partitions.

This was what's causing the error for us earlier.
We have 448 threads with 130 register per thread requested. On paper we have 448 * 130 = 60928 < 65536
but in reality, 130 register per thread is rounded to 136 register per thread (allocation granularity), while 448 threads -> 14 warps, rounded up to 16 warps (512 threads). So hardware check does 512 * 136 = 69632 > 65536 and it's failing the check.

I was copying the checks from `cuda_occupancy.h`, where they were doing the opposite logic.